src/System.Net.Security/src/System/Net/Security/SslStreamInternal.Adapters.cs

```diff
@@ -15,6 +15,39 @@ private interface ISslWriteAdapter
             Task WriteAsync(byte[] buffer, int offset, int count);
         }
 
+        private interface ISslReadAdapter
+        {
+            ValueTask<int> ReadAsync(byte[] buffer, int offset, int count);
+            ValueTask<int> LockAsync(Memory<byte> buffer);
+        }
+
+        private struct SslReadAsync : ISslReadAdapter
+        {
+            private readonly SslState _sslState;
+            private readonly CancellationToken _cancellationToken;
+
+            public SslReadAsync(SslState sslState, CancellationToken cancellationToken)
+            {
+                _cancellationToken = cancellationToken;
+                _sslState = sslState;
+            }
+
+            public ValueTask<int> ReadAsync(byte[] buffer, int offset, int count) => new ValueTask<int>(_sslState.InnerStream.ReadAsync(buffer, offset, count, _cancellationToken));
```

Comment:
true, I wrote this before I was using memory, that valuetask was just to avoid task on the sync method. But you are right I can just avoid the wrapping with memory.